### PR TITLE
Add vulkan backend

### PR DESCRIPTION
**Pull request by kd-11**
1.Most games work, although some colors may be swapped for some texture formats. I'm identifying them and fixing them is trivial. Some random crashes also experienced as well as possible flickering.

2.Window resize does not work very well yet, i'm working on it, should be an easy fix.

3.I've been told that Nvidia driver crashes during intialization, but as I have no access to Nvidia hardware, I'm going to let someone who does find the bug.

This is a port of the OpenGL backend, but the emulation is much better than GL in some games for some reason. Games like afterburner climax work with vulkan but not GL.

### CODE REVIEW

**Comment by vlj** [comment]
Need rebase

File: rpcs3/Emu/RSX/VK/VKFragmentProgram.cpp [hunk 1]
```diff
@@ -0,0 +1,321 @@
...
+
+void VKFragmentProgram::Delete()
+{
+	shader.clear();
+
+	if (handle)
+	{
+		if (Emu.IsStopped())
+		{
+			LOG_WARNING(RSX, "VKFragmentProgram::Delete(): vkDestroyShaderModule(0x%X) avoided", handle);
```
Comment (Nekotekina) [inline rpcs3/Emu/RSX/VK/VKFragmentProgram.cpp]: Oh...

**Comment by MyaniPT** [comment]
AppVeyor failed
  LINK : fatal error LNK1104: cannot open file 'vulkan-1.lib' [C:\rpcs3\rpcs3\rpcs3.vcxproj]

File: rpcs3/Emu/RSX/VK/VKFragmentProgram.cpp (hunk 1 above)
Comment (kd-11) [inline rpcs3/Emu/RSX/VK/VKFragmentProgram.cpp]: Direct OGL port. I forgot how hacky i made some these sections.

**Comment by kd-11** [comment]
What configuration is appveyor building? Release-LLVM should work just fine, although the branch is kinda outdated. I've just plain updated the branch to check if travis and appveyor like the build. Will rebase properly very soon.

File: rpcs3/Emu/RSX/VK/VulkanAPI.h [hunk 2]
```diff
@@ -0,0 +1,11 @@
+#pragma once
+
+#define VK_USE_PLATFORM_WIN32_KHR
```
Comment (danilaml) [inline rpcs3/Emu/RSX/VK/VulkanAPI.h]: what about other OSes? https://github.com/SaschaWillems/Vulkan/blob/master/external/vulkan/vk_platform.h

Comment (vlj) [inline rpcs3/Emu/RSX/VK/VulkanAPI.h]: I guess we can add support later.
I think Vulkan backend still needs works (as every others ones) but it's likely easier if it's in sync with master branch.

Comment (kd-11) [inline rpcs3/Emu/RSX/VK/VulkanAPI.h]: Updated now

Comment (vlj) [inline rpcs3/Emu/RSX/VK/VulkanAPI.h]: BTW currently Vulkan support is really sparse on Linux. It's supported by Skylake gpu (older Intel cpu support is apparently incomplete) on Wayland (Xlib binding was mentionned as untested in the LunarG SDK, although I'm expecting it to work somehow) and with Nvidia with XLib. AMD Vulkan support is Windows only and atm Android support is non existent.

Comment (kd-11) [inline rpcs3/Emu/RSX/VK/VulkanAPI.h]: Seems like it'll take some more work to get it to build on *nix it seems. I forgot about dependancy on some windows stuff that i'm currently using (hInstance, hWnd).

Comment (danilaml) [inline rpcs3/Emu/RSX/VK/VulkanAPI.h]: At least isolate it, so it'd be easier to port it later.

Comment (kd-11) [inline rpcs3/Emu/RSX/VK/VulkanAPI.h]: Ok. I'll isolate vulkan so that it's windows only for now

**Comment by Xcedf** [comment]
Can't build with VS Release-LLVM, same error
LINK : fatal error LNK1104: cannot open file "vulkan-1.lib"

**Comment by kd-11** [comment]
@Xcedf could you check if vulkan-1.lib exists in Vulkan\Vulkan-build\loader\Release? Could be a build order/dependancy issue. vulkan-1.lib is an output of the Vulkan-build project

**Comment by Xcedf** [comment]
Nope, there's only VKstatic.1.lib in the folder

**Comment by vlj** [comment]
If I remove in VKGSRender.h "#pragma comment(lib, "vulkan-1.lib")" it builds and launch with VS Release-llvm.
I don't have the vulkan-1.lib error though, but if I leave the line rpcs3 can't start due to missing vulkan-1.dll.

**Comment by tambry** [comment]
You should be linking with VKstatic.1.lib, that's generated by Vulkan-LoaderAndValidationStuff.

**Comment by tambry** [comment]
Also, what's the speed compared to OGL and DX12 backends?

**Comment by kd-11** [comment]
@tambry GL is currently quite fast. Performance is generally in the same region, but some games are a bit faster with vulkan, and some that are broken on GL work on vulkan even though the codebase is the same. There are advantages since the driver isn't doing checks and guesses to figure out what the emu is trying to do.

**Comment by feliwir** [comment]
From just reading through this i don't understand why you translate the ps3 shaders to glsl and then into spirv, when you could just translate them directly into spirv (since this is standardized aswell)

**Comment by Xcedf** [comment]
Thanks, build success with VKstatic.1.lib instead of vulkan-1.lib

**Comment by kd-11** [comment]
@feliwir I'd also have to learn spir-v and test that bugs in the renderer aren't because of bugs in the shader compiler. Its easier since we know how the glsl compiler behaves. Once we have it running, we'll switch to direct spir-v generation. GLSL is just easier to debug than spir-v.

**Comment by kd-11** [comment]
@vlj is it an instant crash for you when linked statically? i'm assuming you are using nvidia.

**Comment by kd-11** [comment]
I might close this until it's in better shape and compiles properly on linux. Even appveyor does not build, linker error about not finding vkstatic.1.lib.

**Comment by Nekotekina** [comment]
What an urge to close PRs, just keep it opened until it works.

**Comment by vlj** [comment]
@kd-11 yes there's a crash :/ I'm investigating.

**Comment by vlj** [comment]
It looks like nvidia driver is broken with Vulkan, I can't get an instance working

**Comment by tambry** [comment]
@vlj Do validation layers give anything?

**Comment by kd-11** [comment]
@vlj You can quickly enable debug layers here: https://github.com/RPCS3/rpcs3/pull/1548/files#diff-c9e6fd1ef5ae3e53670d088d2961e556R18

**Comment by kd-11** [comment]
Okay, so both appveyor and gcc/clang (linux) are now only complaining about linker error. cannot find vkstatic.1.lib

**Comment by raven02** [comment]
@vlj , are you using this beta driver ? (Vulkan beta driver 356.45)

https://developer.nvidia.com/vulkan-driver

**Comment by vlj** [comment]
I found the issue : 
https://github.com/RPCS3/rpcs3/pull/1548/files#diff-c9e6fd1ef5ae3e53670d088d2961e556R962

apiVersion needs to be build with VK_MAKE_VERSION() macro, ie VK_MAKE_VERSION(1, 0, 0).
With this fix createInstance callback works.

**Comment by kd-11** [comment]
@vlj thats interesting. I inherited that section from some of the early lunarG demos.

**Comment by SakataGintokiYT** [comment]
Can you sent me Vulkan build  for test  if now works on Nvidia Gpu ;)

**Comment by tambry** [comment]
@SakataGintokiYT At least wait until it compiles on AppVeyor. There's probably not much point in testing with it anyways, as the perfomance is only a bit better and it has many issues.

**Comment by MyaniPT** [comment]
AppVeyor isn't building it sadly x.x

**Comment by tambry** [comment]
@kd-11 In rpcs3/CMakeLists.txt try changing _${vulkan}_ to simply _vulkan_. That should work, hopefully.

**Comment by kd-11** [comment]
@tambry direct 64-bit pointer to u32 raises warnings that gcc throws errors on travis-ci

**Comment by kd-11** [comment]
@tambry Will editing cmakelists affect only appveyor?

**Comment by tambry** [comment]
@kd-11 No. CMake is used for both AppVeyor and Travis.

**Comment by vlj** [comment]
It won't change visual studio sln though.

**Comment by SakataGintokiYT** [comment]
Fix crash on Nvidia but not working 

{rsx::thread} RSX: Program destructor invoked!

{rsx::thread} Exception: Assertion failed! Result is C4650338h
(in file Emu\RSX\VK\VKTexture.cpp:162, in function vk::texture::create)

**Comment by kd-11** [comment]
@SakataGintokiYT What causes that error? try running triangle and cube first

**Comment by vlj** [comment]
BTW latest WHQL are compatible with Vulkan according to release note.

**Comment by SakataGintokiYT** [comment]
@kd-11 
Cube log : https://gist.github.com/SakataGintokiYT/b5ce3e9b6c4a9741643c

**Comment by vlj** [comment]
I have the following error from driver debug layer :

·E RSX: ERROR: [Swapchain] Code 16 : vkCreateSwapchainKHR() called with a non-supported pCreateInfo->presentMode (i.e. VK_PRESENT_MODE_IMMEDIATE_KHR).
·W {rsx::thread} RSX: WARNING: [MEM] Code 17 : VkMapMemory: Attempting to map memory range of size zero
·W {rsx::thread} RSX: WARNING: [MEM] Code 17 : Unmapping Memory without memory being mapped: mem obj 0x7108f70
·E {rsx::thread} RSX: ERROR: [PARAMCHECK] Code 1 : vkCreateImage parameter, VkResult result, is unrecognized enumerator

**Comment by kd-11** [comment]
@vlj, could you check with a mail box present mode?

**Comment by kd-11** [comment]
@vlj VK_PRESENT_MODE_FIFO_KHR or K_PRESENT_MODE_MAILBOX_KHR

**Comment by kd-11** [comment]
The last error is because the nvidia driver returns an unknown enum for the error. C4650338h is not defined. The spec only has error enumerators from FFFFFFF5h to FFFFFFFFh

**Comment by vlj** [comment]
swap chain error is fixed using the mailbox mode.

**Comment by kd-11** [comment]
@vlj its interesting that nvidia only supports the vsync modes but unlocked display doesn't work. I'm adding a fix for that soon. Vulkan can be tedious in the number of individual checks needed. I just figured no-vsync would work by default. Does FIFO mode also work?

**Comment by vlj** [comment]
FIFO works too

**Comment by kd-11** [comment]
yeah. I read the spec again. only FIFO is mandated. I'm updating the swapchain generation code.

**Comment by vlj** [comment]
The second error seems to be caused by wrong tiling mode, apparently VK_IMAGE_TILING_OPTIMAL is mandatory in VKTexture.cpp at line 157.

**Comment by kd-11** [comment]
@vlj It shouldn't be mandatory since it means those textures require copying to a staging texture then doing a copyimage call. obviously this isn't ideal. I try to check if linear tiling is supported for a texture format before uploading. The staging texture must have linear tiling enabled so that one can upload texture data manually using the textureutils functions. I'm honestly not sure how to fix that one right now. I'm guessing this happens during staging texture creation, right?

**Comment by vlj** [comment]
Yes it happens at texture creation.
BTW if I use RenderDoc injection in RPCS3 I'm getting a lot more validation error :
https://gist.github.com/vlj/7480ca22a443be5af79c

**Comment by kd-11** [comment]
@vlj Could you post the callstack? I'd like to know why the textures that failed did so. It might just be depth and color attachments or something. I check for tiling support but only for sampled images since they require data to be uploaded.
Also, with renderdoc, i'm assuming that renderdoc is the one injecting those extra descriptors, since the errors are absent without renderdoc even with the same validation layers on. I see the same with official lunarG samples and renderdoc.

**Comment by vlj** [comment]
call stack : https://gist.github.com/vlj/366f774cfa75a937e884

**Comment by kd-11** [comment]
@vlj Thanks. I figured it out

**Comment by kd-11** [comment]
@vlj texture tiling should be fixed now.

**Comment by vlj** [comment]
kd-11 it fixes the error.
Now there is an exception throw in 
"		vram_allocation.allocate_from_pool(device, m_memory_layout.size, m_memory_layout.memoryTypeBits);"
call : "Could not find suitable memory type!"

I think it's because a render target is visible only by the gpu so it shouldn't allocate from a VK_MEMORY_PROPERTY_HOST_VISIBLE_BIT pool.


BTW it's interesting to note that if it works it means that AMD gpu can draw in host visible memory which may lead to some nice optimisation in the long term.

**Comment by kd-11** [comment]
@vlj You're right there. Wow, that NV driver is so restrictive! That does have some consequences when downloading data from color attachments though. Staging textures will be required.

**Comment by vlj** [comment]
I guess that's why dx12 doesn't expose linear tiling texture (except when sharing) and use staging texture for everything.

**Comment by kd-11** [comment]
I've made the render targets gpu only now.

**Comment by vlj** [comment]
Will check tomorrow. 
I don't understand why it doesn't find the lib in appveyor.

**Comment by tambry** [comment]
@vlj Errors caused by injecting RenderDoc are due to false positives detected by validation layers, due to using RenderDoc.

File: rpcs3/CMakeLists.txt [hunk 3]
```diff
@@ -191,14 +191,14 @@ if(WIN32) # I'm not sure we need all of these libs, but we link them in vs
 	else()
 		target_link_libraries(rpcs3  dxgi.lib d2d1.lib dwrite.lib)
 	endif()
-	target_link_libraries(rpcs3  asmjit.lib avformat.lib avcodec.lib avutil.lib swresample.lib swscale.lib png16_static ${wxWidgets_LIBRARIES} ${OPENAL_LIBRARY} ${ADDITIONAL_LIBS} ${vulkan} ${glslang})
+	target_link_libraries(rpcs3  asmjit.lib avformat.lib avcodec.lib avutil.lib swresample.lib swscale.lib png16_static ${wxWidgets_LIBRARIES} ${OPENAL_LIBRARY} ${ADDITIONAL_LIBS} ${vulkan} ${glslang} ${OSDependent} ${OGLCompiler} ${SPIRV})
```
Comment (vlj) [inline rpcs3/CMakeLists.txt]: I think VKstatic.1 target in Vulkan loader project can be used instead of Vulkan target

**Comment by kd-11** [comment]
@vlj OK. I'll try it out tomorrow

**Comment by vlj** [comment]
It still doesn't work with Nvidia but I think we can fix this after merge.

Two errors reported by the debug layers :

"·E {rsx::thread} RSX: ERROR: [MEM] Code 16 : vkCmdBeginRenderPass(): Cannot read invalid memory e6a0b60, please fill the memory before using."
Digging into this error inside the debug layer reveals that depth stencil VK_ATTACHMENT_LOAD_OP_LOAD in renderpass is the culprit.
I think it's because the first time the depth stencil image is created it has undefined content. I'm not sure if it's an issue at driver level or if it's more a warning about API apparent misusage.
Using VK_ATTACHMENT_LOAD_OP_DONT_CARE fix the issue. I think a proper fix would involve issuing a clear command to a rtt surface right after it has been created.

The second error is 
"vkQueuePresentKHR(): Cannot read invalid swapchain image e3f01a0, please fill the memory before using."
This one seems to crash Nvidia driver. Investigating memlayer code didn't bring me a lot of information, I suspect here that the swapchain content is considered undefined as well before being presented but flip() should fill it with the content of surface target A or B.

**Comment by vlj** [comment]
BTW looking at the flip code it looks like there is no barrier to change present image layout.
According to https://www.khronos.org/registry/vulkan/specs/1.0/man/html/vkCmdClearColorImage.html the present image needs to be set in VK_IMAGE_LAYOUT_GENERAL or VK_IMAGE_USAGE_TRANSFER_DST_BIT  layout.

**Comment by kd-11** [comment]
@vlj There is an image barrier during fbo blit. There isnt one during clear though, so i'll add it. The OP_LOAD is there to prevent a draw call from erasing previous draw calls. I dont think there's any way around that, other than just clearing the framebuffer after it is created.

**Comment by kd-11** [comment]
I think i figured out why it crashes during flip for nvidia. Will fix soon.

**Comment by vlj** [comment]
Still crash :/ I'm not sure why.

**Comment by vlj** [comment]
Actually it's better : If I force the "else" path line 1041 in flip() (ie only clear and present) it doesn't crash anymore. I have a black screen of course but flipping does occur (at ~150fps)

**Comment by vlj** [comment]
By the way AMD released a new driver that supports both Vulkan and DX12 : http://support.amd.com/en-us/kb-articles/Pages/AMD_Radeon_Software_Crimson_Edition_16.3.aspx?sf22273025=1

**Comment by kd-11** [comment]
Build should be fixed now for both travis and appveyor. I'll create a new branch to investigate nvidia issues slowly, by starting from a bare implementation and adding features incrementally since I dont own a vulkan-capable nvidia gpu.

**Comment by vlj** [comment]
If it can help the issue with nvidia is in the present mechanism only : I'm using primitive test, and if I use the "clear present image" path I got correct presentation which hints that previous command are correctly working.

It looks like nvogl-64 is trying to access an invalid pointer in present call.

**Comment by kd-11** [comment]
@vlj The present mechanism is a little messy, some parts are queued and some aren't. Try https://github.com/kd-11/rpcs3/tree/vulkan-nvidia-tests and see if it flips. If not, i'll try to rewrite the lunarg demos to use offscreen rendering and see if it still causes problems during flip. Might be a sync issue or a driver bug. I'm certain i'm probably missing something very small somewhere. Vulkan really needs a reference implementation.

**Comment by vlj** [comment]
I found what's wrong : there is no wait on m_submit_fence after submission here https://github.com/RPCS3/rpcs3/pull/1548/files#diff-bb1c9be33467ff75ef0fcca2e9c849fbR1053 

It looks like there is a concurrent access done on the present_queue by the submit command call and present one.

**Comment by kd-11** [comment]
I did suspect a silly sync issue. Does waiting on the fence fix it?

**Comment by vlj** [comment]
yes, I added a 			CHECK_RESULT(vkWaitForFences((*m_device), 1, &m_submit_fence, VK_TRUE, ~0ULL)); and it doesn't crash anymore.

**Comment by kd-11** [comment]
The wait is actually here https://github.com/RPCS3/rpcs3/pull/1548/files#diff-bb1c9be33467ff75ef0fcca2e9c849fbR1000. I added some stuff later and forgot to move it. :p

**Comment by MyaniPT** [comment]
{rsx::thread} Exception: Unknown blend op: 0x8008
(in file C:\rpcs3\rpcs3\Emu\RSX\VK\VKGSRender.cpp:154, in function vk::get_blend_op)
screenshot_126

Works on DX12 and OpenGL

**Comment by kd-11** [comment]
@MyaniPT You dont have LLE enabled for cellSail and cellSailAvi for resogun.

**Comment by MyaniPT** [comment]
@kd-11 LLE'd cellSpursJq , cellSre, cellSail and cellSailAvi
screenshot_127

**Comment by kd-11** [comment]
@MyaniPT Alot of pipeline state switches arent implemented yet, just a handful of blend modes and other basic stuff. Alot of fixes will be coming in the next few weeks after this is merged.

**Comment by kd-11** [comment]
I just noticed that this PR messes up wxWidgets so dont merge just yet

**Comment by kd-11** [comment]
Fixed wxWidgets issue.

**Comment by danilaml** [comment]
@kd-11 Maybe it's time for a big clean rebase?

**Comment by kd-11** [comment]
@danilaml I'm not quite sure. The commits touch on different little things e.g cmake configurations, main gui settings, etc. If something breaks, it may be difficult to single out the cause. If its too much however, I could rebase it to a handful of commits.

**Comment by kd-11** [comment]
Rebased to 14 commits

**Comment by danilaml** [comment]
I think last one (https://github.com/kd-11/rpcs3/commit/9ce9eef9f8b1975464c0f9bd039e6b198d01297e) can be (and should be if no problems arise) squashed/fix up with https://github.com/kd-11/rpcs3/commit/403b9e3d1332baece78e4282285540a1513b6fe9

**Comment by kd-11** [comment]
Appveyor needs restarting.

**Comment by kd-11** [comment]
Rebased and reworded the commits.

**Comment by vlj** [comment]
I don't know how to restart build :/

**Comment by MyaniPT** [comment]
AppVeyor failed due to a timeout while fetching a submodule. Try making another commit and it should start building a new build

**Comment by danilaml** [comment]
@vlj I've restarted it. If you are not part of appveyor rpcs3 team ask @tambry

**Comment by vlj** [comment]
thanks, merged

**Pull request merged**